extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-preview.js

```diff
@@ -0,0 +1,654 @@
+    deviceSpecs.element = deviceLayout;
+    this.element.appendChild(deviceLayout);
+    deviceSpecs.chip = this.addDeviceChip_(deviceSpecs);
+    this.devices_.push(deviceSpecs);
+    this.onLayoutMeasure();
+    setStyles(deviceLayout, {'opacity': '1'});
+    setTimeout(() => {
+      setStyles(deviceLayout, {
+        'transition': 'transform ease-out 0.15s',
+      });
+    }, 150);
+  }
+
+  /**
+   * @private
+   * @param {DeviceInfo} device
+   */
+  removeDevice_(device) {
+    device.chip.setAttribute('inactive', '');
+    setTimeout(() => {
+      device.chip.remove();
+    }, 200);
+    this.devices_ = this.devices_.filter((d) => d != device);
+    device.element.remove();
+    this.onLayoutMeasure();
+  }
+
+  /**
+   * @param {DeviceInfo} device
+   * @return {!Element} the device chip
+   */
+  addDeviceChip_(device) {
+    const deviceChip = buildDeviceChipTemplate(this.element);
+    deviceChip.querySelector('span').textContent = device.name;
+    deviceChip.device = device;
+    this.element
+      .querySelector('.i-amphtml-story-dev-tools-device-chips')
+      .appendChild(deviceChip);
+    return deviceChip;
+  }
+
+  /**
+   * @private
+   */
+  updateDevicesInHash_() {
+    const devicesStringRepresentation = this.devices_
+      .map((device) => {
+        if (device.custom) {
+          return (
+            `${device.width}:${device.height}` +
+            (device.name ? ':' + device.name : '')
+          );
+        }
+        return device.name.toLowerCase().replace(/[^a-z0-9]/gi, '');
+      })
+      .join(';');
+    this.element.setAttribute('devices', devicesStringRepresentation);
+    updateHash({'devices': devicesStringRepresentation}, this.win);
+  }
+
+  /** @override */
+  onLayoutMeasure() {
+    const layoutBox = this.getLayoutBox();
+    let sumDeviceWidths = 0;
+    let maxDeviceHeights = 0;
+    this.devices_.forEach((deviceSpecs) => {
+      sumDeviceWidths += deviceSpecs.width;
+      maxDeviceHeights = Math.max(
+        maxDeviceHeights,
+        deviceSpecs.deviceHeight || deviceSpecs.height + 100
+      );
+    });
+    const scale = Math.min(
+      (layoutBox.width / sumDeviceWidths) * 0.9,
+      (layoutBox.height / maxDeviceHeights) * 0.8
+    );
+    let cumWidthSum = 0;
+    const paddingSize =
+      (layoutBox.width - sumDeviceWidths * scale) / (this.devices_.length + 1);
+    this.devices_.forEach((deviceSpecs, i) => {
+      const scaleWidthChange = deviceSpecs.width * (scale - 1) * 0.5; // Accounts for width change on scaling
+      const cumPaddings = (i + 1) * paddingSize;
+      const leftOffset = cumPaddings + cumWidthSum + scaleWidthChange;
+      setStyles(deviceSpecs.element, {
+        'transform': `perspective(100px) translateZ(${
+          (100 * (scale - 1)) / scale
+        }px) translateX(${leftOffset / scale}px)`,
+      });
+      cumWidthSum += deviceSpecs.width * scale;
+    });
+  }
+
+  /**
+   * @private
+   */
+  showAddDevicePopup_() {
```

Comment:
This method does DOM read/write all over the place. Please batch DOM reading / DOM writing and wrap them in measureElement (if needed, only for things that trigger a relayout) and mutateElement.
ALso, please use event delegation and come up with an overall better pattern to do so all over this file.

Reply:
Added the dialog on a mutateElement because editing the element when it's not attached shouldn't trigger any changes. And improved the event delegation overall in the whole element!